src/services/offer.js

```diff
@@ -192,29 +196,108 @@ class OfferService {
+            offers = this._buildInitialSearchQuery(queryValue, showHidden, showAdminReason, queryFilters);
+        }
 
-        const offersQuery = offers
-            .sort(value ? { score: { "$meta": "textScore" } } : undefined)
-            .skip(offset)
+        const results = await offers
+            .sort(queryValue ? { score: { "$meta": "textScore" }, _id: 1 } : { _id: 1 })
             .limit(limit)
         ;
 
-        return showAdminReason ? offersQuery : offersQuery.select("-adminReason");
+        if (results.length > 0) {
+            const lastOffer = results[results.length - 1];
+            return {
+                results,
+                queryToken: this.encodeQueryToken(
+                    lastOffer._id,
+                    lastOffer.score || lastOffer._doc?.score,
+                    queryValue, queryFilters
+                ),
+            };
+        } else {
+            return { results };
+        }
+    }
+
+    /**
+     * Builds an initial search query. Cannot be used when loading more offers.
+     * Otherwise, use _buildSearchContinuationQuery().
+     */
+    _buildInitialSearchQuery(value, showHidden, showAdminReason, filters) {
+        const offers = (value ? Offer.find({ "$and": [
+            this._buildFilterQuery(filters),
+            { "$text": { "$search": value } }
+        ] }, { score: { "$meta": "textScore" } }
+
+        ) : Offer.find(this._buildFilterQuery(filters)));
+
+        offers.current();
+        if (!showHidden) offers.withoutHidden();
+        if (!showAdminReason) offers.select("-adminReason");
+
+        return offers;
+    }
+
+    /**
+     * Builds a search continuation query. Only use this when loading more offers.
+     * Otherwise, use _buildInitialSearchQuery().
+     */
+    _buildSearchContinuationQuery(lastOfferId, lastOfferScore, value, showHidden, showAdminReason, filters) {
+        let offers;
+        if (value) {
+            offers = Offer.aggregate([
+                { $match: { $text: { $search: value } } },
+                { $match: this._buildFilterQuery(filters) },
+                { $addFields: {
+                    score: { $meta: "textScore" },
+                    adminReason: { $cond: [showAdminReason, "$adminReason", "$$REMOVE"] }
+                } },
+                { $match: { "$or": [
+                    { score: { "$lt": lastOfferScore } },
+                    { score: lastOfferScore, _id: { "$gt": ObjectId(lastOfferId) } }
+                ] } },
+                { $match: Offer.filterCurrent() },
+                { $match: showHidden ? {} : Offer.filterNonHidden() }
+            ]);
+        } else {
+            offers = Offer.find({ "$and": [
+                this._buildFilterQuery(filters),
+                { _id: { "$gt": ObjectId(lastOfferId) } }
+            ] });
+
+            offers.current();
+            if (!showHidden) offers.withoutHidden();
+            if (!showAdminReason) offers.select("-adminReason");
```

Comment:
Same here for using the same offers object every time

e.g. `offers = offers.current()`, etc...

Reply:
Same answer as above